### PR TITLE
[ME-129] Add correct answers and pry rails gem

### PR DESCRIPTION
**Pull request by PSkoczylas**
Add correct answers field to ```/api/tests/{uuid}/scores```.

### CODE REVIEW

File: tests-api/spec/use_cases/calculate_score_spec.rb [hunk 1]
```diff
@@ -128,8 +128,9 @@
     end
     let(:score) { "#{answers.size}/#{answers.size}" }
     let(:results) do
-      test.questions.map { |q| { question_uuid: q.id, correct: true } }
+      test.questions.map { |q| { question_uuid: q.id, correct_answers: expected_answers, correct: true } }
     end
+    let(:expected_answers) { answers.map { |answer| answer[:selected_choices][0] } }
```
Comment (aitorlb) [inline tests-api/spec/use_cases/calculate_score_spec.rb:133]: since the logic is getting complicated, I encourage you to write a tests for when `result` contains correct and incorrect answers
 
```ruby
let(:results) do
  [ { question_uuid: "question_uuid", correct_answers: ["whatever"], correct: true }]  # context with correct submitted answers
end

let(:results) do
  [ { question_uuid: "question_uuid", correct_answers: ["whatever"], correct: false }]  # context with incorrect submitted answers
end

File: tests-api/app/use_cases/calculate_score.rb [hunk 2]
```diff
@@ -31,23 +31,28 @@ def calculate_results
...
+        correct_choices,
       )
 
-      @correct_questions += 1 if correct
+      @correct_questions += 1
 
       {
         question_uuid: answer[:question_uuid],
+        correct_answers: correct_choices,
         correct: correct,
```
Comment (aitorlb) [inline tests-api/app/use_cases/calculate_score.rb:45]: maybe rename `correct` to avoid confusion with `correct_answers` ? `answered_correctly` maybe?

File: tests-api/app/use_cases/calculate_score.rb [hunk 3]
```diff
@@ -31,23 +31,28 @@ def calculate_results
...
 
     answers.map do |answer|
+      correct_choices = correct_choices(answer[:question_uuid])
       correct = answered_correctly?(
-        answer[:question_uuid],
         answer[:selected_choices],
+        correct_choices,
       )
 
-      @correct_questions += 1 if correct
```
Comment (aitorlb) [inline tests-api/app/use_cases/calculate_score.rb:39]: this looks like a bug, if you write a test for an incorrect submitted answer, you will catch it (or not).

File: tests-api/app/use_cases/calculate_score.rb [hunk 4]
```diff
@@ -31,23 +31,28 @@ def calculate_results
...
 
-  def answered_correctly?(question_uuid, selected_choices)
-    question        = questions.find(question_uuid)
-    correct_choices = question.choices.select { |c| c[:correct] }.map { |c| c[:answer] }
+  def correct_choices(question_uuid)
+    question = questions.find(question_uuid)
+    question.choices.select { |c| c[:correct] }.map { |c| c[:answer] }
+  end
+
+  def answered_correctly?(selected_choices, correct_choices)
```
Comment (aitorlb) [inline tests-api/app/use_cases/calculate_score.rb:55]: maybe switch the parameters order? this way we follow the standard of `actual` vs `expectation`, which is reflected in the body of the method:
```ruby
(correct_choices - selected_choices).empty?
```

File: tests-api/spec/use_cases/calculate_score_spec.rb [hunk 5]
```diff
@@ -126,11 +126,37 @@
         results: results,
       }
     end
-    let(:score) { "#{answers.size}/#{answers.size}" }
-    let(:results) do
-      test.questions.map { |q| { question_uuid: q.id, correct: true } }
+
```
Comment (aitorlb) [inline tests-api/spec/use_cases/calculate_score_spec.rb:129]: how about something like this to remove duplication?
```ruby
describe "#data" do
  subject(:data) { described_class.call(params: params).data }

  # params
  let(:answers) do
    [{ question_uuid: question_uuid, selected_choices: selected_choices }]
  end

  # data
  let(:expected_data) do
    { score: score, results: results }
  end
  let(:results) do
    [{ question_uuid: question_uuid, correct_answers: correct_answers, correct: correct }]
  end

  let(:question) { test.questions.first }
  let(:question_uuid) { question.id }

  let(:correct_answers) do
    question.choices.select { |c| c[:correct] }.map { |c| c[:answer] }
  end

  describe "when results includes correct submitted answers" do
    let(:score) { "1/1" }
    let(:selected_choices) { correct_answers }
    let(:correct) { true }
    # ...

  describe "when results includes incorrect submitted answers" do
    let(:score) { "0/1" }
    let(:selected_choices) { ["Incorrect answer"] }
    let(:correct) { false }
    # ...
```

Comment (PSkoczylas) [reply tests-api/spec/use_cases/calculate_score_spec.rb:129]: Added some of such changes.

File: tests-api/app/use_cases/calculate_score.rb [hunk 6]
```diff
@@ -31,23 +31,28 @@ def calculate_results
...
       )
 
-      @correct_questions += 1 if correct
+      @correct_questions += 1 if correct_answer
 
       {
         question_uuid: answer[:question_uuid],
-        correct: correct,
+        correct_answers: correct_choices,
+        correct: correct_answer,
```
Comment (aitorlb) [inline tests-api/app/use_cases/calculate_score.rb:45]: > maybe rename correct to avoid confusion with correct_answers ? answered_correctly maybe?

I was thinking about the public API here, to make it more understandable for the outside world. WDYT?

```ruby
{
  score: "1/1",
  results: [
    {
      question_uuid: "uuid",
      correct_answers: ["a", "b", "c"],
      answered_correctly: true, # instead of :correct
    },
  ],
}
```

File: tests-api/spec/swagger_helper.rb [hunk 7]
```diff
@@ -16,10 +16,10 @@
   # the root example_group in your specs, e.g. describe '...', swagger_doc: 'v2/swagger.json'
   config.swagger_docs = {
     "v1/swagger.yaml" => {
-      swagger: "2.0",
+      openapi: "3.0.1",
```
Comment (aitorlb) [inline tests-api/spec/swagger_helper.rb:19]: why this change?

Comment (aitorlb) [reply tests-api/spec/swagger_helper.rb:19]: lesson-api uses [swagger 2.0](https://lesson-api-test.herokuapp.com/swagger/?format=openapi)

File: tests-api/app/use_cases/calculate_score.rb [hunk 8]
```diff
@@ -31,23 +31,28 @@ def calculate_results
...
-      correct = answered_correctly?(
-        answer[:question_uuid],
+      correct_choices = correct_choices(answer[:question_uuid])
+      correct_answer = answered_correctly?(
+        correct_choices,
         answer[:selected_choices],
       )
 
-      @correct_questions += 1 if correct
+      @correct_questions += 1 if correct_answer
```
Comment (aitorlb) [inline tests-api/app/use_cases/calculate_score.rb:40]: maybe `@correct_questions` should be renamed to `@correct_answers`

Comment (aitorlb) [reply tests-api/app/use_cases/calculate_score.rb:40]: or maybe not :thinking:
